src/utp_wrapper.rs

```diff
@@ -32,15 +37,40 @@ impl UtpWrapper {
                 .spawn(move || {
                     let mut socket = socket;
                     socket.set_read_timeout(Some(CHECK_FOR_NEW_WRITES_INTERVAL_MS));
+                    let heartbeat_msg = unwrap_result!(serialise(&CrustMsg::Heartbeat));
+                    let mut heartbeat_deadline = Instant::now() + heartbeat_timeout;
+                    let mut inactivity_deadline = Instant::now() + inactivity_timeout;
+                    let update_heartbeat = |heartbeat_deadline: &mut Instant, inactivity_deadline: &mut Instant| {
+                        let now = Instant::now();
+                        *heartbeat_deadline = now + heartbeat_timeout;
+                        *inactivity_deadline = now + inactivity_timeout;
+                    };
                     'read_write: loop {
                         let mut buf = [0; BUFFER_SIZE];
+                        {
+                            let now = Instant::now();
+                            if now > inactivity_deadline {
+                                info!("Stale connection. Dropping...");
+                                break;
+                            }
+                            if now > heartbeat_deadline {
+                                if let Err(e) = socket.send_to(&heartbeat_msg) {
+                                    error!("Error sending: {:?}", e);
+                                    break;
+                                }
+                                update_heartbeat(&mut heartbeat_deadline,
+                                                 &mut inactivity_deadline);
```

Comment:
> But the implementation should only update `inactivity_deadline` when it receives data from the peer, not whenever it sends a heartbeat.

But the process of sending hearbeats will force both peers to exchange messages (ack and stuff). Sending messages is blocking and signalizes errors through the return value, so I assume both peers communicated after `send` returns.

> Also the `heartbeat_deadline` shouldn't be updated every time we receive data from the peer otherwise, if they're sending us data, we'll never send them a heartbeat and they may end up thinking we've disconnected.

But we'll send them an ACK instead a heartbeat. We acknowledge every data we receive.